drivers/flash/soc_flash_nrf5.c

```diff
@@ -161,12 +199,357 @@ static const struct flash_driver_api flash_nrf5_api = {
+{
+	int result;
+	u8_t ticker_id;
+	u8_t instance_index;
+	struct flash_op_desc *op_desc;
+
+	if (ll_radio_state_is_idle()) {
+		op_desc = context;
+
+		if (op_desc->handler(op_desc->context)
+			== FLASH_OP_DONE) {
+			ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+			/* Stop the time slot ticker */
+			result = ticker_stop(instance_index,
+					     0,
+					     ticker_id,
+					     NULL,
+					     NULL);
+
+			if ((result != TICKER_STATUS_SUCCESS)
+			    && (result != TICKER_STATUS_BUSY)) {
+				__ASSERT(0, "Failed to stop ticker.\n");
+			}
+
+			((struct flash_op_desc *)context)->result = 0;
+
+			/* notify thread that data is available */
+			k_sem_give(&sem_sync);
+		}
+	} else {
+		__ASSERT(0, "Radio is on during flash operation.\n");
+	}
+}
+
+static void time_slot_callback_helper(u32_t ticks_at_expire, u32_t remainder,
+		u16_t lazy, void *context)
+{
+	u8_t instance_index;
+	u8_t ticker_id;
+	int err;
+
+	ll_radio_state_abort();
+
+	ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+	/* start a secondary one-shot ticker after ~ 500 us, */
+	/* this will let any radio role to gracefully release the Radio h/w */
+
+	err = ticker_start(instance_index, /* Radio instance ticker */
+		0, /* user_id */
+		0, /* ticker_id */
+		ticks_at_expire, /* current tick */
+		TICKER_US_TO_TICKS(500), /* first int. */
+		0, /* periodic (on-shot) */
+		0, /* per. remaind. (on-shot) */
+		0, /* lazy, voluntary skips */
+		0,
+		time_slot_callback_work, /* handler for executing */
+					 /* the flash operation */
+		context, /* the context for the flash operation */
+		NULL, /* no op callback */
+		NULL);
+
+	if ((err != TICKER_STATUS_SUCCESS) && (err != TICKER_STATUS_BUSY)) {
+		((struct flash_op_desc *)context)->result = -ECANCELED;
+
+		/* abort flash timeslots */
+		err = ticker_stop(instance_index, 0, ticker_id, NULL, NULL);
+
+		if ((err != TICKER_STATUS_SUCCESS)
+			&& (err != TICKER_STATUS_BUSY)) {
+			__ASSERT(0, "Failed to stop ticker %d.\n");
+		}
+
+		/* notify thread that data is available */
+		k_sem_give(&sem_sync);
+	}
+}
+
+
+static int work_in_time_slice(struct flash_op_desc *p_flash_op_desc)
+{
+	u8_t instance_index;
+	u8_t ticker_id;
+	u32_t err;
+
+	ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+	err = ticker_start(instance_index,
+			   3, /* user id for thread mode */
+			      /* (MAYFLY_CALL_ID_PROGRAM) */
+			   ticker_id, /* flash ticker id */
+			   ticker_ticks_now_get(), /* current tick */
+			   0, /* first int. immediately */
+			   TICKER_US_TO_TICKS(FLASH_INTERVAL), /* periodic */
+			   TICKER_REMAINDER(FLASH_INTERVAL), /* per. remaind.*/
+			   0, /* lazy, voluntary skips */
+			   TICKER_US_TO_TICKS(FLASH_SLOT),
+			   time_slot_callback_helper,
+			   p_flash_op_desc,
+			   NULL, /* no op callback */
+			   NULL);
+
+	int result;
+
+	if ((err != TICKER_STATUS_SUCCESS) && (err != TICKER_STATUS_BUSY)) {
+		result = -ECANCELED;
+	} else if (k_sem_take(&sem_sync, K_MSEC(200)) != 0) {
```

Comment:
No! It is time for waiting on procedure done - if it wouldn't happened the timeout would occur. Indeed the limit should be much longer (for full chip-erase) or even unlimited.